SDWebImage/SDWebImageManager.m

```diff
@@ -340,20 +346,34 @@ - (void)callCompletionBlockForOperation:(nullable SDWebImageCombinedOperation*)o
 @end
 
 
+@interface SDWebImageCombinedOperation ()
+
+@property (strong, nonatomic, nonnull) dispatch_semaphore_t cancelLock; // a lock to make the `cancel` method thread-safe
```

Comment:
😅 Using a semaphore here...Well it's OK. But since most cases, the `cancel` method is called only once, there is no different performance impact for it. Using `@synchronized` is enough.

Reply:
But it's a bit better and no harm done, right? 😅

Reply:
The advantage to use `@synchronized` because it's easy, and process many race condition during using. For example, `@synchronized` is a recursive lock, means if you are using a `@synchronized(self)`, it's safe to again using `@synchronized(operation)` outside of your own code and avoid deadlock, which semaphore can not do it simply. And also, it add a try-catch blcok to unlock the lock when a exception was through. So even a runtime exception occur, the lock itself will not cause a unfail deadlock...

I guess the proper way to using semaphore, is when using with `NSMutableArray`, `NSMutableDictionary` collection. You already know that this mutable property is not cause recursive lock, and no exception will occur, so it's safe and maybe more performent to use this. For a single object which method may be called only once, I guess `@synchronized` is better...

Reply:
Wow, That's thoughtful about the `@synchronized`. I've benefited from it, Thanks :)

I will revert it.